File: profile-form-FE/src/compareProfiles.js
function compareProfiles(p1, p2) {
  // Destructring
  let {
    gender: gender1,
    major: major1,
    personality: personality1,
    romance: romance1,
    friendship: friendship1,
    hobbies: hobbies1,
    spirituality: spirituality1,
    partying: partying1
  } = p1;
  if (gender1 === "Male") {
    gender1 = "Men";
  }
  if (gender1 === "Female") {
    gender1 = "Women";
  }
  let {
    gender: gender2,
    major: major2,
    personality: personality2,
    romance: romance2,
    friendship: friendship2,
    hobbies: hobbies2,
    spirituality: spirituality2,
    partying: partying2
  } = p2;
  if (gender2 === "Male") {
    gender2 = "Men";
  }
  if (gender2 === "Female") {
    gender2 = "Women";
  }

  // Score to be returned
  let score = 0;

  // Not looking for the same thing --> Fail Case
  if (
    romance1.includes("Not looking for this") &&
    friendship2.includes("Not looking for this")
  ) {
    return 0;
  }
  if (
    romance2.includes("Not looking for this") &&
    friendship1.includes("Not looking for this")
  ) {
    return 0;
  }

  // Gender Mismatch --> Fail Case
  if (
    !romance1.includes("Not looking for this") ||
    !romance2.includes("Not looking for this")
  ) {
    if (!romance2.includes(gender1) || !romance1.includes(gender2)) {
      return 0;
    }
  } else if (
    !friendship1.includes("Not looking for this") ||
    !friendship2.includes("Not looking for this")
  ) {
    if (!friendship2.includes(gender1) || !friendship1.includes(gender2)) {
      return 0;
    }
  }

  // Major (10% weight out of 100points)
  if (major1 === major2) {
    score += 10;
  }

  // Personality (10% weight out of 100points)
  if (personality1 === personality2) {
    score += 10;
  }

  // Hobbies (50% weight out of 100points)
  let hobbyPoints = 50 / Math.min(hobbies1.length, hobbies2.length);
  let moreHobby = hobbies1.length > hobbies2.length ? hobbies1 : hobbies2;
  let lessHobby = hobbies1.length <= hobbies2.length ? hobbies1 : hobbies2;

  for (let hobby in lessHobby) {
    if (!moreHobby.includes(hobby)) {
      score += hobbyPoints;
    }
  }

  // Spirituality (10% weight out of 100points)
  if (spirituality1 === spirituality2) {
    score += 10;
  }

  // Party Favors (20% weight out of 100points)
  let partyPoints = 20 / Math.min(partying1.length, partying2.length);
  let moreParty = partying1.length > partying2.length ? partying1 : partying2;
  let lessParty = partying1.length <= partying2.length ? partying1 : partying2;

  for (let party in lessParty) {
    if (!moreParty.includes(party)) {
      score += partyPoints;
    }
  }

  // Return Score
  return score;
}

export default compareProfiles;

File: backend/compareProfiles.py
def compareProfiles(p1, p2):

    score = 0

    if (p1['gender'] == "Male"):
        p1['gender'] = "Men"
    if (p1['gender'] == "Female"):
        p1['gender'] = "Women"

    if (p2['gender'] == "Male"):
        p2['gender'] = "Men"
    if (p2['gender'] == "Female"):
        p2['gender'] = "Women"

    # Not looking for the same thing --> Fail Case
    if (
        "Not looking for this" in p1['romance'] and
        "Not looking for this" in p2['friendship']
    ):
        return 0

    if (
        "Not looking for this" in p2['romance'] and
        "Not looking for this" in p1['friendship']
    ):
        return 0

    # Gender Mismatch --> Fail Case
    if (
        "Not looking for this" in p1['romance'] or
        "Not looking for this" in p2['romance']
    ):
        if (p1['gender'] not in p2['friendship'] or p2['gender'] not in p1['friendship']):
            return 0

    if (
        "Not looking for this" in p1['friendship'] or
        "Not looking for this" in p2['friendship']
    ):
        if (p1['gender'] not in p2['romance'] or p2['gender'] not in p1['romance']):
            return 0

    # Major (10% weight out of 100points)
    if (p1['major'] == p2['major']):
        score += 10

    # Personality (10% weight out of 100points)
    if (p1['personality'] == p2['personality']):
        score += 10

    # Hobbies (50% weight out of 100points)
    hobbyPoints = 50 / min(len(p1['hobbies']), len(p2['hobbies']))
    moreHobby = p1['hobbies'] if (
        len(p1['hobbies']) > len(p2['hobbies'])) else p2['hobbies']
    lessHobby = p1['hobbies'] if (
        len(p1['hobbies']) <= len(p2['hobbies'])) else p2['hobbies']

    for hobby in lessHobby:
        if (hobby in moreHobby):
            score += hobbyPoints

    # Spirituality (10% weight out of 100points)
    if (p1['spirituality'] == p2['spirituality']):
        score += 10

    # Party Favors (20% weight out of 100points)
    partyPoints = 20 / min(len(p1['partying']), len(p2['partying']))
    moreParty = p1['partying'] if (
        len(p1['partying']) > len(p2['partying'])) else p2['partying']
    lessParty = p1['partying'] if (
        len(p1['partying']) <= len(p2['partying'])) else p2['partying']

    for party in lessParty:
        if (party in moreParty):
            score += partyPoints

    # Return Score
    return score
